[WPE] Enforce surface size match when maximized to avoid Wayland errors

psaavedra opened the pull request:
#### 93908c419dcdad47fd745a43b126fd539274ec27
<pre>
[WPE] Enforce surface size match when maximized to avoid Wayland errors
<a href="https://bugs.webkit.org/show_bug.cgi?id=293172">https://bugs.webkit.org/show_bug.cgi?id=293172</a>

Reviewed by Carlos Garcia Campos.

Prevents xdg_wm_base.invalid_surface_state protocol violations by
checking if view is in maximized state before rendering buffer and
skipping buffers with mismatched scaled dimensions.

This is based on fix proposed in Cog[1].

[1] <a href="https://github.com/Igalia/cog/pull/539">https://github.com/Igalia/cog/pull/539</a>

* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandRenderBuffer):

Canonical link: <a href="https://commits.webkit.org/295091@main">https://commits.webkit.org/295091@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/042bfb2989b1982425dc67964c171e499ea6921f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104085 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109281 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54753 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24156 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32333 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/79074 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107091 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18763 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/93892 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59401 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18566 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54113 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88286 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11995 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111667 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31241 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23032 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88084 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31605 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90086 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87741 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32623 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10372 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25695 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16893 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31170 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36482 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30964 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34300 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32524 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -358,6 +358,11 @@ static gboolean wpeViewWaylandRenderBuffer(WPEView* view, WPEBuffer* buffer, con
     if (!wlBuffer)
         return FALSE;
 
+    if (wpe_view_get_width(view) != wpe_buffer_get_width(buffer) || wpe_view_get_height(view) != wpe_buffer_get_height(buffer)) {
```
You can't compare view and buffer size, because the view size is in logical coords.

psaavedra replied:
Could you elaborate a bit more on the explanation? I don’t quite understand the real difference between logical coordinates and the buffer size. The empirical tests I’ve done suggest that they are indeed comparable. Aren’t both measured in _pixels_? Could it be because the window values are affected by a `scale` factor? Can you suggest an alternative?

carlosgcampos replied:
It works if device scale factor is 1 only because coords match in that case. The buffer size is scaled by device scale factor, while the view size isn't.

psaavedra replied:
So if  the patch has in consideration the scale factor I understand it will be right. Right?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -358,6 +358,11 @@ static gboolean wpeViewWaylandRenderBuffer(WPEView* view, WPEBuffer* buffer, con
         return FALSE;
 
+    if (wpe_view_get_width(view) != wpe_buffer_get_width(buffer) || wpe_view_get_height(view) != wpe_buffer_get_height(buffer)) {
+        wpe_view_buffer_rendered(view, buffer);
```
This might be problematic too, we are expected to reply right after the next vblank.

psaavedra replied:
While it's true that you're correct in theory, this particular case is very specific. Window resizing is an occasional event that happens quite rarely and indeed the user it can expect some discontinuity in the frame rate sync.

To meet your expectations, I’m thinking we could keep calling to `wpe_view_buffer_rendered()` inside a `wl_surface_frame()` listener but avoid attaching the buffer to the `surface` using `wl_surface_attach()`. That means something like _do nothing but wait for the next tick_. 

That would preserve the frame throttling hint. However, it seems to me like overcomplicating things for a sporadic situation.

carlosgcampos replied:
I see, while the window is being resized we can receive a buffer for the previous size. In that case it's probably ok to just skip immediately since we have already requested another one for the current size. The question now is, why is this specific to wayland? I wonder if should handle this at AcceleratedBackingStoreDMABuf and simply not call wpe_view_render_buffer()

carlosgcampos replied:
And call frameDone() of course, to notify the web process

psaavedra replied:
My understanding is Wayland (more specific) the Wayland compositor (in this case Weston) performs a check in the server side. Unfortunately this check is not just a warning but aborts all the rendering. I don't know how others backends deal with this situation. Are you thinking in GTK? I don't know to be honest.

carlosgcampos reviewed (changes requested):
In which situation can this happen? Maybe we should try to avoid getting to this point earlier instead.

psaavedra committed:
[WPE] Enforce surface size match when maximized to avoid Wayland errors
https://bugs.webkit.org/show_bug.cgi?id=293172

Reviewed by Carlos Garcia Campos.

Prevents xdg_wm_base.invalid_surface_state protocol violations by
checking if view is in maximized state before rendering buffer and
skipping buffers with mismatched scaled dimensions.

This is based on fix proposed in Cog[1].

[1] https://github.com/Igalia/cog/pull/539

* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandRenderBuffer):

Canonical link: https://commits.webkit.org/295091@main
